src/main/java/com/laytonsmith/core/events/drivers/PlayerEvents.java

```diff
@@ -857,57 +888,53 @@ public void bind(BoundEvent event) {
 
 		@Override
 		public boolean matches(Map<String, Mixed> prefilter, BindableEvent e) throws PrefilterNonMatchException {
-			if(e instanceof MCPlayerInteractEvent) {
-				MCPlayerInteractEvent pie = (MCPlayerInteractEvent) e;
+			throw new UnsupportedOperationException();
+		}
 
-				if(prefilter.containsKey("button")) {
-					if(pie.getAction().equals(MCAction.LEFT_CLICK_AIR)
-							|| pie.getAction().equals(MCAction.LEFT_CLICK_BLOCK)) {
-						if(!prefilter.get("button").val().toLowerCase().equals("left")) {
-							return false;
+		@Override
+		public PrefilterBuilder<MCPlayerInteractEvent> getPrefilters() {
+			return new PrefilterBuilder<MCPlayerInteractEvent>()
+					.set("button", "\"left\" or \"right\". If they left or right clicked",
+							new StringICPrefilterMatcher<>() {
+						@Override
+						protected String getProperty(MCPlayerInteractEvent pie) {
+							if(pie.getAction().equals(MCAction.LEFT_CLICK_AIR)
+								|| pie.getAction().equals(MCAction.LEFT_CLICK_BLOCK)) {
+								return "left";
+							}
+							if(pie.getAction().equals(MCAction.RIGHT_CLICK_AIR)
+									|| pie.getAction().equals(MCAction.RIGHT_CLICK_BLOCK)) {
+								return "right";
+							}
+							throw new Error("Unexpected event behavior, please report this bug to developers.");
 						}
-					}
-					if(pie.getAction().equals(MCAction.RIGHT_CLICK_AIR)
-							|| pie.getAction().equals(MCAction.RIGHT_CLICK_BLOCK)) {
-						if(!prefilter.get("button").val().toLowerCase().equals("right")) {
-							return false;
+					})
+					.set("itemname", "The item type they are holding when they interacted, or null",
+							new MaterialPrefilterMatcher<>() {
+						@Override
+						public MCMaterial getMaterial(MCPlayerInteractEvent pie) {
+							return pie.getItem().getType();
 						}
-					}
-				}
-
-				if(prefilter.containsKey("itemname")) {
-					Mixed item = prefilter.get("itemname");
-					MCMaterial mat = pie.getItem().getType();
-					if(mat == null) {
-						if(!(item instanceof CNull)) {
-							return false;
+					})
+					.set("block", "The block type the player interacts with, or null if nothing",
+							new MaterialPrefilterMatcher<>() {
+						@Override
+						public MCMaterial getMaterial(MCPlayerInteractEvent pie) {
+							return pie.getClickedBlock().getType();
```

Comment:
I'm tempted to make these subclasses for MCItemStack and MCBlock, so the null handling logic can live in only one place.

Reply:
I'm not exactly sure what you mean, but keep in mind that whether an event can return null for an item or block (etc) is specific to the event. So writing it in the event's prefilter should be fine.

Reply:
I mean that if the block or item stack is null, then the material would also be null, which would then be compared against CNull. The check would look the same - if the method would never return null, then the if(block == null) { return null; } would never happen anyways, but in the events where it can happen, it wouldn't NPE, it would just return null.